File: 2001-3000/2813-maximum-elegance-of-a-k-length-subsequence/2813-maximum-elegance-of-a-k-length-subsequence.java
class Solution {
    public long findMaximumElegance(int[][] arr, int k) {
        int n = arr.length;
        long s = 0, ms = 0, distinct = 0;
        int[] freq = new int[n + 1];
        Arrays.sort(arr, (a, b) -> b[0] - a[0]);
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        for (int i = 0; i < k; i++) {
            s += arr[i][0];
            if (freq[arr[i][1]] == 0)
                distinct++;
            if (freq[arr[i][1]] > 0)
                pq.add(new int[] { arr[i][0], arr[i][1] });
            freq[arr[i][1]]++;
            ms = Math.max(s + distinct * distinct, ms);

        }
        for (int i = k; i < n; i++) {
            if (freq[arr[i][1]] == 0) {
                if (!pq.isEmpty()) {
                    s -= pq.poll()[0];
                    s += arr[i][0];
                    distinct++;
                    freq[arr[i][1]]--;
                    ms = Math.max(s + (distinct) * (distinct), ms);
                }
            }
        }
        return ms;
    }
}

File: 2001-3000/2813-maximum-elegance-of-a-k-length-subsequence/2813-maximum-elegance-of-a-k-length-subsequence.cpp
class Solution {
public:
    using ll = long long;
    long long findMaximumElegance(vector<vector<int>>& items, int k) {
        ll sum = 0, ret = 0, cnt = 0;
        ranges::sort(items);
        int n = items.size();
        vector<int> freq(n + 1);
        for(int i = n - 1; i >= n - k; --i)
        {
            sum += items[i][0];
            if(freq[items[i][1]]++ == 0)
            {
                sum += (2 * (cnt++) + 1);
            }
        }
        ret = sum;
        for(int i = n - k, j = n - k - 1; i < n && j >= 0; ++i)
        {
            if(freq[items[i][1]] == 1) continue;
            while(j >= 0 && freq[items[j][1]] > 0)
            {
                --j;
            }
            if(j < 0) break;
            ++freq[items[j][1]];
            --freq[items[i][1]];
            sum += items[j][0] - items[i][0] + (2 * (cnt++) + 1);
            ret = max(ret, sum);
            --j;
        }
        return ret;
    }
};